sign/ed448/ed448.go

```diff
@@ -0,0 +1,187 @@
+// previously-generated private key.
+func NewKeyFromSeed(seed PrivateKey) *KeyPair {
+	if len(seed) != Size {
+		panic("ed448: bad private key length")
+	}
+	var h [2 * Size]byte
+	H := sha3.NewShake256()
+	_, _ = H.Write(seed)
+	_, _ = H.Read(h[:])
+	clamp(h[:Size])
+
+	var _r goldilocks.Scalar
+	_r.FromBytes(h[:Size])
+	pk := &KeyPair{}
+	goldilocks.Curve{}.ScalarBaseMult(&_r).ToBytes(pk.public[:])
+	copy(pk.private[:], seed)
+	return pk
+}
+
+// Sign returns the signature of a message using both the private and public
+// keys of the signer.
+func Sign(kp *KeyPair, message, context []byte) []byte {
+	if len(context) > 255 {
+		panic("context should be at most 255 octets")
+	}
+	var r, h, hRAM [2 * Size]byte
+	H := sha3.NewShake256()
+	_, _ = H.Write(kp.private[:])
+	_, _ = H.Read(h[:])
+
+	prefix := [10]byte{'S', 'i', 'g', 'E', 'd', '4', '4', '8', byte(0), byte(len(context))}
+	H.Reset()
+	_, _ = H.Write(prefix[:])
+	_, _ = H.Write(context)
+	_, _ = H.Write(h[Size:])
+	_, _ = H.Write(message)
+	_, _ = H.Read(r[:])
+
+	var _r goldilocks.Scalar
+	_r.FromBytes(r[:])
+	signature := make([]byte, 2*Size)
+	goldilocks.Curve{}.ScalarBaseMult(&_r).ToBytes(signature[:Size])
+
+	H.Reset()
+	_, _ = H.Write(prefix[:])
+	_, _ = H.Write(context)
+	_, _ = H.Write(signature[:Size])
+	_, _ = H.Write(kp.public[:])
+	_, _ = H.Write(message)
+	_, _ = H.Read(hRAM[:])
+
+	var _hRAM, _h, _s goldilocks.Scalar
+	_hRAM.FromBytes(hRAM[:])
+	clamp(h[:Size])
+	_h.FromBytes(h[:Size])
+	_s.Mul(&_h, &_hRAM)
+	_s.Add(&_s, &_r)
+	copy(signature[Size:], _s[:])
+	return signature
+}
+
+// Verify returns true if the signature is valid. Failure cases are invalid
+// signature, or when the public key cannot be decoded.
+func Verify(public PublicKey, message, context, signature []byte) bool {
+	P, errDecoding := goldilocks.FromBytes(public)
+	if len(public) != Size ||
+		len(signature) != 2*Size ||
+		len(context) > 255 ||
+		!isLessThanOrder(signature[Size:]) ||
+		errDecoding != nil {
+		return false
+	}
+
+	var hRAM [2 * Size]byte
+	prefix := [10]byte{'S', 'i', 'g', 'E', 'd', '4', '4', '8', byte(0), byte(len(context))}
+	H := sha3.NewShake256()
+	_, _ = H.Write(prefix[:])
+	_, _ = H.Write(context)
+	_, _ = H.Write(signature[:Size])
+	_, _ = H.Write(public[:Size])
+	_, _ = H.Write(message)
+	_, _ = H.Read(hRAM[:])
+
+	var _s, _hRAM goldilocks.Scalar
+	_hRAM.FromBytes(hRAM[:])
+	_s.FromBytes(signature[Size:])
+
+	P.Neg()
+	var enc [Size]byte
+	goldilocks.Curve{}.CombinedMult(&_s, &_hRAM, P).ToBytes(enc[:])
+	return bytes.Equal(enc[:], signature[:Size])
+}
+
+// isLessThanOrder returns true if 0 <= x < order.
+func isLessThanOrder(x []byte) bool {
+	order := goldilocks.Curve{}.Order()
+	i := len(order) - 1
+	for i > 0 && x[i] == order[i] {
+		i--
+	}
+	return x[Size-1] == 0 && x[i] < order[i]
+}
+
+func clamp(k []byte) {
+	k[0] &= 252
+	k[Size-2] |= 0x80
+	k[Size-1] = 0x00
+}
+
+func makeCopy(in *[Size]byte) []byte {
+	out := make([]byte, Size)
+	copy(out, in[:])
```

Comment:
This might be good to do in constant time as it is sometimes copying the private key...

Reply:
does copy present time variability?

Reply:
I think it's constant time, but in this case we can just write

```go
ret := *in
return ret[:]
```

which is definitely constant time.  Note that it is **not** the same as

```go
return (*in)[:]
```

Reply:
I agree with @bwesterb ;)